File: utils/cache.go
// Package cache implements a cache on top of a map (dict/hashtable)
// A Least Recently Used (LRU) list of keys is maintained, so that when
// the cache is about to exceed its maximum capacity, it can evict the
// LRU item and recover its space.
package cache

import (
	"container/list" // used to implement the LRU list, O(1)
	"fmt"
)

// A single item in the cache.
type cache_item struct {
	headers    map[string]string // header map of key: value
	file_bytes []byte            // file bytes
	item_size  int               // memory used by this item
}

// Members of the cache 'class'.
type cache struct {
	keys          map[string]cache_item // map of URL: item
	max_bytes     int
	current_bytes int
	LRU           *list.List // Least Recently Used list of keys
}

// Construct a new cache and return it.
func New(max_bytes int) cache {
	keys := make(map[string]cache_item)
	LRU := list.New()
	c := cache{keys, max_bytes, 0, LRU}
	return c
}

// Print the cache contents.
func (c *cache) Print() {
	multiple := "s"
	if 1 == c.LRU.Len() {
		multiple = ""
	}
	fmt.Printf("%d Cache Item%s:\n", c.LRU.Len(), multiple)
	for key, item := range c.keys {
		fmt.Printf("  %s\n", key)
		fmt.Println("    headers:", item.headers)
		fmt.Println("    file len:", len(item.file_bytes))
	}
	fmt.Println("LRU list (last key is LRU):")
	for e := c.LRU.Front(); e != nil; e = e.Next() {
		fmt.Println("  ", e.Value)
	}
	fmt.Println("Max cache size:", c.max_bytes, "bytes")
	fmt.Println("  Current size:", c.current_bytes, "bytes")
	fmt.Println("        Unused:", c.max_bytes-c.current_bytes, "bytes")
}

// Set a header into the cache
func (c *cache) SetHeader(key string, subkey string, value string) {
	c.checkSizeAndEvict(len(value))        // Do we need to make room?
	if item, found := c.keys[key]; found { // key exists, so add subkey
		item.headers[subkey] = value
		item.item_size += len(value)
	} else { // key not found, so add it
		headers := make(map[string]string)
		headers[subkey] = value
		item := cache_item{headers: headers, item_size: len(value)}
		c.keys[key] = item
	}
	c.current_bytes += len(value) // Total bytes in the cache
	// Put this key in the front (most recently used) spot in the LRU list.
	c.addHeadToLRU(key)
}

// Set a file into the cache
func (c *cache) SetFile(key string, value []byte) {
	c.checkSizeAndEvict(len(value))        // Do we need to make room?
	if item, found := c.keys[key]; found { // key exists, so set value
		item.file_bytes = value
		item.item_size += len(value)
		c.keys[key] = item // replace item with updated one
	} else { // key not found, so add it
		headers := make(map[string]string)
		item := cache_item{headers: headers, file_bytes: value,
			item_size: len(value)}
		c.keys[key] = item
	}
	c.current_bytes += len(value) // Total bytes in the cache
	// Put this key in the front (most recently used) spot in the LRU list.
	c.addHeadToLRU(key)
}

// Get a header by key from the cache.
func (c *cache) GetHeader(key string, subkey string) string {
	c.addHeadToLRU(key) // Move this key to the front of the LRU list.
	return c.keys[key].headers[subkey]
}

// Get a file by key from the cache.
func (c *cache) GetFile(key string) []byte {
	c.addHeadToLRU(key) // Move this key to the front of the LRU list.
	return c.keys[key].file_bytes
}

// If the key is in the list, remove it.
// This function is not exported, so like private, since it starts with a
// lower case letter.
func (c *cache) removeFromLRU(key string) {
	// (The linear search below will be slow if many items in the list.)
	for i := c.LRU.Front(); i != nil; i = i.Next() {
		if key == i.Value {
			c.LRU.Remove(i)
			return
		}
	}
}

// Put this key at the head of the list (make it the Most recently used).
func (c *cache) addHeadToLRU(key string) {
	// If the key is already in the list in another position, remove it.
	c.removeFromLRU(key)
	// Put the key at the front of the list.
	c.LRU.PushFront(key)
}

// Returns the key of the LRU item, or an empty string if the LRU list is empty.
func (c *cache) getLRU() string {
	if 0 == c.LRU.Len() {
		return ""
	}
	i := c.LRU.Back() // last item in list, so LRU
	ret := i.Value
	c.LRU.Remove(i)     // remove the LRU item
	return ret.(string) // convert List Element interface to string
}

// Will adding this value exceed our our capacity?
// If so, evict the LRU item to make room.
// Arguments: the size of the item to add.
func (c *cache) checkSizeAndEvict(value_size int) {
	if c.current_bytes+value_size < c.max_bytes {
		return // Enough space in the cache, so nothing to do.
	}

	// Capacity will be exceeded, so get and evict the LRU
	lru := c.getLRU()
	if "" == lru {
		return // The list/cache is empty
	}

	// Recover the bytes used by this item
	recovered_bytes := c.keys[lru].item_size
	c.current_bytes -= recovered_bytes
	if 0 > c.current_bytes {
		c.current_bytes = 0
	}

	// Remove the LRU from the cache dict
	delete(c.keys, lru) // delete is a built in, works on maps.

	// Remove the LRU from the list
	c.removeFromLRU(lru)
	fmt.Println("Evicted", lru, "from the cache and recovered", recovered_bytes,
		"bytes.")
}

File: utils/cache.py
#!/usr/bin/env python3.6

from collections import deque
from typing import Dict, Any


#------------------------------------------------------------------------------
class Cache:
    ''' Helper class to cache HTTP headers and file contents from an URL.
        A Least Recently Used (LRU) list of URLs is maintained, so that when 
        the cache is about to exceed its maximum capacity, it can evict the
        LRU item and recover its space.
    '''
    # Content cache in a hashtable.  Assumes a single host.
    # Format is {'URL', {'header': 'value', 
    #                    'file_bytes': <file contents>}}
    __cache: Dict[str, Dict[str, Any]] = {}


    def __init__(self, max_bytes: int = 200*1024) -> None:
        self.__LRU = deque() # Least Recently Used list of keys
        self.max_bytes = max_bytes
        self.current_bytes = 0


    def __repr__(self) -> str:
        ret = ''
        for k in self.__cache.keys():
            ret += k + '\n'
            for sk in self.__cache[k].items():
                if sk[1] is not None and len(sk[1]) <= 70: 
                    ret += f'  {sk[0]}: {sk[1]}\n'
                else:
                    ret += f'  {sk[0]}: ...\n' # don't print large values
        ret += f'LRU list (last key is LRU):\n'
        for k in self.__LRU:
            ret += f'  {k}\n'
        ret += f'Max cache size: {self.max_bytes} bytes\n'
        ret += f'  Current size: {self.current_bytes} bytes\n'
        ret += f'        Unused: {self.max_bytes - self.current_bytes} bytes\n'
        multiple = 's'
        if 1 == len(self.__LRU):
            multiple = ''
        return f'{len(self.__LRU)} Cache Item{multiple}:\n{ret}'


    def clear(self) -> None:
        self.__cache.clear()
        self.__LRU.clear()
        self.current_bytes = 0


    def set(self, key: str, subkey: str, value: Any) -> None:
        if value is None:
            return
        # Make room for this item if necessary.
        self.__check_size_and_evict(value)
        # Add the new item to the cache
        if key not in self.__cache: # add the first entry
            self.__cache[key] = {subkey: value}
        else: 
            if subkey in self.__cache[key]:
                # we are over-writing a subkey, so recover its bytes
                self.current_bytes -= len(self.__cache[key][subkey])
            self.__cache[key][subkey] = value
        # Account for the memory this value takes up (ignore keys)
        self.current_bytes += len(value)
        # Put this key in the front (most recently used) spot in the LRU list.
        self.__add_head_to_LRU(key)


    def get(self, key: str, subkey: str) -> Any:
        if key not in self.__cache: 
            return None
        elif subkey not in self.__cache[key]:
            return None
        # Move this key to the front of the LRU list.
        self.__add_head_to_LRU(key)
        return self.__cache[key][subkey]


    def __remove_from_LRU(self, key: str) -> None:
        # If the key is in the list, remove it.
        # (The linear search below will be slow if many items in the list.)
        if key in self.__LRU:      
            self.__LRU.remove(key)  


    def __add_head_to_LRU(self, key: str) -> None:
        # Remove the key if it's already in the list.
        self.__remove_from_LRU(key)
        # Put the key at the front of the list.
        self.__LRU.appendleft(key)


    def __get_LRU(self) -> str:
        # Returns the key of the LRU item, or None if the LRU list is empty.
        if 0 == len(self.__LRU):
            return None
        return self.__LRU.pop() # Note: removes item from the list.


    # Will adding this value exceed our our capacity?
    # If so, evict the LRU item to make room.
    def __check_size_and_evict(self, value_to_add: Any) -> None:
        if self.current_bytes + len(value_to_add) < self.max_bytes:
            return # Nothing to do, there is enough space.
        # Capacity will be exceeded, so get and evict the LRU
        lru = self.__get_LRU()
        if lru is None:
            return
        recovered_bytes = 0
        # Recover the bytes from the subkeys of this item
        for sk in self.__cache[lru].items():
            recovered_bytes += len(sk[1])
        self.current_bytes -= recovered_bytes
        # Remove the LRU from the cache dict
        self.__cache.pop(lru)
        # Remove the LRU from the list
        self.__remove_from_LRU(lru)
        print(f'Evicted {lru} from the cache and recovered '
                f'{recovered_bytes} bytes.') 

